ci: use windows 2025 runners

ShortDevelopment opened the pull request:
`windows-2019` actions are unsupported as of 30.06.2025.

This PR updates the ci-config to run on the supported versions of the windows runner:
- `windows-2022` → `windows-2025`
- `windows-2019` → `windows-2022`

## 💥 Breaking changes
- The default CC configuration on windows used the [WinGlob](https://learn.microsoft.com/en-us/uwp/api/windows.globalization) api.
  This config is broken on recent versions of Windows 11 as CC does not initialize WinRT using [RoInitialize](https://learn.microsoft.com/en-us/windows/win32/api/roapi/nf-roapi-roinitialize) to allow for some optimizations.
- CC now uses the version of libICU that is [shipped with Windows](https://learn.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-) by default
  This removes compat of the default config with all windows versions < Version 1903 (May 2019 Update)
  Should be fine as Windows 10 is out-of-support anyway...

## 🧪 Tests
- Tests did assume `win32 => WinGlob`; that is obviously wrong now.
- Some baselines were WinGlob only => updated
- libICU did return incorrect values so two tests were disabled => fixed
  The "new" behavior is consistent with recent node.js and bun.js
  - #4490
  - #2964 

---

See https://github.com/actions/runner-images/issues/12045
Fixes #7045
Fixes #4490

ShortDevelopment committed:
`windows-2022` → `windows-2025`
`windows-2019` → `windows-2022`

ShortDevelopment committed:
Use windows icu by default

rhuanjl commented:
So, the Windows globalisation implementation doesn't work on latest windows AND the ICU alternative fails too, weirdly the test runner seems to be crashing out after one of the fails rather than finishing and saying how many tests failed.... that is awkward.

A couple of test fails look like things that are may be fairly easily fixable e.g. ICU version mismatches to amend tests for or inconsistent conditions in the test suite that expect windows to not be ICU etc. Hopefully it's not overall too big a job but I have no setup I can work on this on.

ShortDevelopment commented:
I'll look into this in the next view weeks

ShortDevelopment committed:
Link against "new" `icu.lib` on windows.

https://learn.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-#version-1903-may-2019-update

ShortDevelopment committed:
Fix `Intl.getCanonicalLocales` tests

ShortDevelopment committed:
Check icu version dynamically for windows

ShortDevelopment committed:
Assume windows version will always be >= 62

rhuanjl commented:
The windows-2019 test runners are now disabled.... So all CI runs now fail.

ShortDevelopment committed:
Update `ES6_intl_simple_attach.js` baseline

ShortDevelopment commented on Build/Chakra.Build.Default.props:
```diff
@@ -29,10 +29,10 @@
     <BuildChakraICUData Condition="'$(BuildChakraICUData)'=='' AND ('$(ChakraICU)'=='static' OR '$(ChakraICU)'=='shared')">true</BuildChakraICUData>
 
-    <IcuLibraryDependencies Condition="'$(ChakraICU)'=='windows'">icuuc.lib;icuin.lib</IcuLibraryDependencies>
+    <IcuLibraryDependencies Condition="'$(ChakraICU)'=='windows'">icu.lib</IcuLibraryDependencies>
```
See [History of changes to the ICU library in Windows](https://learn.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-#history-of-changes-to-the-icu-library-in-windows).

rhuanjl replied:
Would it be possible to do version detection and build with the older libs if someone is building on older windows? I know we can't test it but may be helpful to someone.

ShortDevelopment replied:
If you decide to support this config I could add another build-flag or detect the config based on the current windows version.

rhuanjl replied:
@ppenzin what do you think of this?

Previously we used Windows Globalisation on windows for the support we use ICU for on Linux and macOS.

- I have long hoped we could standardise on ICU to avoid this platform difference.
- A recent windows update broke CC's use of Windows Globalisation, we could fix it but it seemed an opportune time to instead switch windows to ICU.
- BUT from May 2019 onwards windows has shipped with a single DLL for ICU before that it was two DLLs if we build expecting the single DLL (as per this PR) we don't support those older (unsupported) versions of windows

ppenzin replied:
Would it be a big change to try to detect if one or two DLLs needed? I would lean towards having the check, but if additional logic makes configuration too complicated, I this we should be OK breaking older versions.

ShortDevelopment replied:
Adding a check would be very easy using the [`Exists` condition](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditions) in MSBuild. I would simply check if the newer `icu.dll` exists in `System32` and use that instead of the older two if available.

The actual problem is that we might start using apis in the future that won't be available in either of the older two libs.
And we won't get any ci failures for that as we're not running ci on such old windows version.
(This PR is because win-2019 is no longer supported on Azure after all)

ShortDevelopment replied:
I'm now checking for the existence of `icu.dll` in the `System32` folder of the machine we're building on (See e661393116151cb5dff7c361003002fefef34902).
This checks if we are buildin _on_ a machine that supports the "new" libicu not whether we compile _against_ a version of windows, that supports it.

ShortDevelopment commented on test/DebuggerCommon/ES6_intl_simple_attach.js.dbg.baseline:
```diff
@@ -131,7 +131,7 @@
               "constructor": "function <large string>",
               "resolvedOptions": "function <large string>",
-              "compare": "function <large string>"
+              "compare": "Error <large string>"
```
I'm not sure why these results changed...

rhuanjl replied:
Are we running this on linux/macOS as well?

I.e. is the windows change bringing windows in line with the other platforms?

ShortDevelopment replied:
Yes, this test was WinGlob-only but now is running on all platforms.
This baseline is generated by the debugger so there might be an issue with that?

ShortDevelopment committed:
Remove `require_winglob`

ShortDevelopment committed:
Fix copyright

ShortDevelopment committed:
Remove bi-di chars from `IntlReturnedValueTests.js` baseline

rhuanjl commented:
@ppenzin What do you think? This fixes INTL on windows newer than the May 2019 Update but breaks it on older versions - though the key change is just build config we could perhaps document how to build the older version?

rhuanjl commented on test/Intl/GetCanonicalLocales.js:
```diff
@@ -90,18 +91,24 @@ var tests = [
-            equal("de-DE-u-co-kn", "de-DE-u-co-yes-kn-true", gcl("de-de-u-kn-co")[0]);
-            equal("de-DE-u-co-phonebk-kn", "de-DE-u-co-phonebk-kn-true", gcl("de-de-u-kn-co-phonebk")[0]);
-            equal("de-DE-u-co-phonebk-kn-yes", "de-DE-u-co-phonebk-kn-true", gcl("de-DE-u-kn-yes-co-phonebk")[0]);
+            if (WScript.Platform.ICU_VERSION < 62) {
```
Why are we only testing this on old versions of ICU? Is this behaviour wrong or broken on new ICU?

Same point below.

ShortDevelopment replied:
It was actually broken in old icu version and behaving as expected using WinGlob.
See #4490

rhuanjl replied:
Is this being used as an always false condition to avoid testing broken behaviour?

ShortDevelopment replied:
The correct behavior is tested in the else branch.

Before these tests used a custom `equals` function that tested the correct behavior for WinGlob and the wrong behavior for libicu. I guess this is why the `exclude_icu62AndAboveTestFailures` was added.
I still kept the old tests as CC technically still supports this version of libicu.

I deduce the condition from the flag name (`exclude_icu62AndAboveTestFailures`); so it maybe should be `<= 62`?

rhuanjl commented on test/Intl/IntlReturnedValueTests.js.dbg.baseline:
```diff
@@ -21,14 +21,14 @@
     "functionCallsReturn": {
       "[get format returned]": "function <large string>",
-      "[Intl.DateTimeFormat.prototype.format returned]": "string ‎2‎/‎1‎/‎2000"
+      "[Intl.DateTimeFormat.prototype.format returned]": "string 2/1/2000"
```
What has changed here?

ShortDevelopment replied:
The output contained [bi-di markers](https://www.unicode.org/reports/tr9/) which normally aren't used in locales like `en-US`.

<img width="980" height="87" alt="image" src="https://github.com/user-attachments/assets/64f79dbf-2b6e-4628-a150-7efb496915d3" />

This also seemed to be a WinGlob quirk:
https://github.com/chakra-core/ChakraCore/blob/792ee7660cbb14f1b3f9f77a6f919aeed4ec1fc2/test/Intl/rlexe.xml#L84-L88

ppenzin replied:
I think we've ran into mysterious mismatch like this before

rhuanjl commented on test/CMakeLists.txt:
```diff
@@ -8,8 +8,6 @@ endif ()
 if (NO_ICU)
     set(TEST_ICU --not-tag exclude_noicu)
-elseif (NOT EMBED_ICU)
-    set(TEST_ICU --not-tag exclude_icu62AndAboveTestFailures)
```
Have you removed all uses of this flag?

ShortDevelopment replied:
Yes. This flag was only used for `NumberFormat.js` and `GetCanonicalLocales.js` (two of the tests that were failing initially)

rhuanjl commented on lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp:
```diff
@@ -2437,7 +2437,7 @@ DEFINE_ISXLOCALEAVAILABLE(PR, uloc)
         }
 
-#if defined(ICU_VERSION) && ICU_VERSION >= 61
+#if !defined(ICU_VERSION) || ICU_VERSION >= 61
```
Is this condition ever going to be false now?

ShortDevelopment replied:
This code guards the usage of apis that were not available before `icu < 61`.
Windows icu does not expose a compile-time `ICU_VERSION` as the exact version is dependent on the windows version at runtime. Instead available apis are determined based on the windows version CC is compiled against.
https://github.com/chakra-core/ChakraCore/blob/792ee7660cbb14f1b3f9f77a6f919aeed4ec1fc2/lib/Runtime/PlatformAgnostic/ChakraICU.h#L9-L18
The idea is: Run the new code on `icu >= 61` if we know the version at built-time else assume the newer apis are available (on windows).

rhuanjl replied:
So if we build on windows ICU_VERSION will be undefined? And therefore this condition will be true?

Please could you add a comment to explain that as it's a little counter intuitive. (Same for the equivalent condition below)

ShortDevelopment replied:
> So if we build on windows ICU_VERSION will be undefined? And therefore this condition will be true?

Exactely

> Please could you add a comment to explain that as it's a little counter intuitive. (Same for the equivalent condition below)

Sure: 1684f20f08c6bc552079fb6a672a162107b521ef

rhuanjl reviewed (commented):
Thank you for the work on this. Hoping we'll merge it within a week though I have a few questions - see comments.

ShortDevelopment committed:
Add comment to explain `ICU_VERSION` conditional-compilation.

ppenzin reviewed (approved):
LGTM; I am fine with either direction on detecting two DLLs vs one - it would be good to detect it, but not sure if additional logic is worthwhile.

ShortDevelopment committed:
Check for existence of `icu.dll` in system32

ppenzin reviewed (approved):
LGTM :shipit:
